### ISSUE TITLE
Live data source doesn't work

### ISSUE THREAD

**Issue by peternewman** [k4kfh/casparcg-html-templates#2]
> I have known SimpleWeather is broken for a while, and IIRC, I had a proof of concept using OpenWeather's free tier a while back, but I haven't had much time to devote to this project the last few months. IIRC, the reason I didn't convert this to OpenWeather is because you'd have to be on the paid tier to get a 7 day forecast. Since I wrote this for a public school, there is no way they can justify the paid tier of OpenWeather. I agree that this needs addressing, however, and dummy data may be the best way to do it. Give me a few days to look this over, I'm at Uni right now so things are sort of nuts for me.

_Originally posted by @k4kfh in https://github.com/k4kfh/casparcg-html-templates/pull/1#issuecomment-529235194_

> Update: perhaps we could do some mild refactoring to use this: https://github.com/BrookeDot/SimplerWeather 1000 free calls a day is more than enough for me, so that solves my problem.

_Originally posted by @k4kfh in https://github.com/k4kfh/casparcg-html-templates/pull/1#issuecomment-529240996_

**Comment by k4kfh** [comment]
After looking at it a little, I am leaning towards Open Weather Map. They are clearly a big company and not going anywhere anytime soon. I don't know as much about Dark Sky, but mostly I don't want to rely on another library for something so simple. With OWM, I don't even need a library, because I'm only making one call. It's not as "programmatic", but it's so much simpler, and that's really what this needs to be, IMO.

**Comment by peternewman** [comment]
So I looked at a few others:

- Yahoo Weather API (requires registration, although I think they basically all do)
- Dark Sky - 1000 calls/day, 7 day forecast - (needs a server side proxy - https://darksky.net/dev/docs/faq#cross-origin), attribution required ( https://darksky.net/dev/docs/faq#attribution )
- US National Weather Service - https://www.weather.gov/documentation/services-web-api - no registration, but US only (which doesn't work for me), also looks like it would be more effort to extract the data, no easy n day forecast
- UK Met Office - https://metoffice.apiconnect.ibmcloud.com/metoffice/production/ - 360 calls/day, 10,000/month - global
- Weatherbit.io - 500 calls/day, 16 day forecast - https://www.weatherbit.io/pricing#api_tiers - [Non-commercial Use](http://creativecommons.org/licenses/by-nc/4.0/)
- AccuWeather - 50 calls/day, 5 day forecast - https://developer.accuweather.com/packages
- OpenWeatherMap - 60 calls/minute, 5 day forecast - https://openweathermap.org/price

**Comment by peternewman** [comment]
> After looking at it a little, I am leaning towards Open Weather Map.

See my summary above, the benefit of DarkSky is a longer forecast, the downside is needing something server side. Weatherbit could work (which has the best of most worlds), depending on if your usage counts as non-commercial.

I guess the ideal will be if it's coded in such a way you can drop in different providers to suit, but that's probably a whole other library!

For bonus points, make it scale the tiles dynamically based on the number of days the forecast provides...

**Comment by k4kfh** [comment]
> See my summary above, the benefit of DarkSky is a longer forecast, the downside is needing something server side. Weatherbit could work (which has the best of most worlds), depending on if your usage counts as non-commercial.

That's a negative for me on DarkSky then. I don't need a longer forecast, and since this is going in a public school (and will be maintained and used by non-programmers) I really don't need another complex piece of equipment to maintain. It's bad enough trying to teach them Caspar, lol.

> For bonus points, make it scale the tiles dynamically based on the number of days the forecast provides...

I would like to create a set of optional parameters (with defaults, but settable via JSON parameters from the CCG client) that control things like number of forecast days, etc. For my use case, they almost always need 7 days, but I can see the utility for other people.

**Comment by peternewman** [comment]
> > See my summary above, the benefit of DarkSky is a longer forecast, the downside is needing something server side. Weatherbit could work (which has the best of most worlds), depending on if your usage counts as non-commercial.
> 
> That's a negative for me on DarkSky then. I don't need a longer forecast, and since this is going in a public school (and will be maintained and used by non-programmers) I really don't need another complex piece of equipment to maintain. It's bad enough trying to teach them Caspar, lol.

Yeah likewise for my particular use case too. Actually it looks like the CORS tests may be turned off in the Chromium within Caspar:
https://github.com/CasparCG/server/blob/6ee169018e7bfd935bb8330683733b430a08bd7f/src/modules/html/producer/html_producer.cpp#L381

So it would only be an issue for development! Attribution would still need covering though.

> > For bonus points, make it scale the tiles dynamically based on the number of days the forecast provides...
> 
> I would like to create a set of optional parameters (with defaults, but settable via JSON parameters from the CCG client) that control things like number of forecast days, etc. For my use case, they almost always need 7 days, but I can see the utility for other people.

I probably didn't explain my point very well, Open Weather Map's free tier only includes a five day forecast ( https://openweathermap.org/price ), so ideally if that is used, and only five days returned, the system should hide the last two tiles and increase the width of the others to suit.

Yeah the options thing was my thinking for temperature units too.

**Comment by k4kfh** [comment]
> I probably didn't explain my point very well, Open Weather Map's free tier only includes a five day forecast ( https://openweathermap.org/price ), so ideally if that is used, and only five days returned, the system should hide the last two tiles and increase the width of the others to suit.

Yeah, that makes more sense. I'll see what I can do. I'm thinking I'll also try to incorporate some error screens (e.g. "You didn't specify your ZIP code!" and similar). I know that's not strictly "professional", but since this is used by non-programmers in an elementary school, it's a convenient way to reduce text messages and phone calls to me asking what's going on.

**Comment by peternewman** [comment]
How do you normally deploy it? Who does which bits?

If you enter the URL and the users just add the JSON, then we could add master options URL encoded which can be overridden by the config JSON. Obviously I'd prefer to avoid messages appearing if possible, unless you're in debug mode, in which case they can be useful. I'd hope they can be made to toggle somehow, but out of the control of users.

There's also console.log, which I believe appears in the Caspar log, but it's not quite as user-friendly.

**Comment by k4kfh** [comment]
At my schools, we use the generic CasparCG Client (as opposed to some of
the ninja CG operators who make their own clients). If you're curious, we
use it in conjunction with a Blackmagic ATEM, although that has little
relevance to this software. We add the template to the Client rundown, then
in the options, choose "Send as JSON", then fill in our ZIP code there. All
the configuration is done in the options.

I understand in a "real" broadcast application, it would be highly
undesirable to show any kind of back end error information on the output.
However, for a group of 6th graders, that's the most straightforward way to
communicate with them and their teacher easily (and the world won't stop
turning if there are technical difficulties  and mistakes on these
broadcasts).

The logical solution, to me, is to make this "Userspace Errors" feature
correspond to a config value. A "default" options configuration
(overridable by the JSON options sent from CCG Client) could be included in
the JavaScript, and in this repo it could have UserspaceErrors:true, but it
would take all of 5 seconds to change the default options to say
UserspaceErrors:false. This is what I would like to do because it covers my
use case (this repo serves as a sort of central download point for my
schools if they need a fresh copy of the template), but makes it trivial
for someone else to customize the default behavior to fit their needs.
